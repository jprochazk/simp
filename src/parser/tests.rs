use std::fmt::Write as _;
use std::path::Path;

fn parse(code: &str) -> String {
    match super::parse(code) {
        Ok(ast) => {
            // we want slightly smaller snapshots, so debug-fmt the contents directly
            let mut o = String::new();
            for stmt in &ast.body {
                writeln!(&mut o, "{stmt:#?}").ok();
            }
            if let Some(tail) = &ast.tail {
                writeln!(&mut o, "{tail:#?}").ok();
            }
            o
        }
        Err(err) => format!("{}", err.render(code)),
    }
}

#[glob_test::glob("./inputs/**/*.simp")]
fn test_parse(path: &Path) {
    let input = std::fs::read_to_string(path).unwrap();
    insta::assert_snapshot!(parse(&input));
}
