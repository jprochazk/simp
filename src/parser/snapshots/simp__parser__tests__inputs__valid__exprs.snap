---
source: src/parser/tests.rs
expression: parse(&input)
---
ExprBinary {
    lhs: ExprIdent {
        name: "or",
    },
    op: Or,
    rhs: ExprBinary {
        lhs: ExprBinary {
            lhs: ExprIdent {
                name: "and0",
            },
            op: And,
            rhs: ExprIdent {
                name: "and0",
            },
        },
        op: And,
        rhs: ExprIdent {
            name: "and1",
        },
    },
}
ExprBinary {
    lhs: ExprIdent {
        name: "and",
    },
    op: And,
    rhs: ExprBinary {
        lhs: ExprBinary {
            lhs: ExprIdent {
                name: "eq0",
            },
            op: Equal,
            rhs: ExprIdent {
                name: "eq0",
            },
        },
        op: Equal,
        rhs: ExprIdent {
            name: "eq1",
        },
    },
}
ExprBinary {
    lhs: ExprIdent {
        name: "eq",
    },
    op: Equal,
    rhs: ExprBinary {
        lhs: ExprBinary {
            lhs: ExprIdent {
                name: "ord0",
            },
            op: LessThan,
            rhs: ExprIdent {
                name: "ord0",
            },
        },
        op: LessThan,
        rhs: ExprIdent {
            name: "ord1",
        },
    },
}
ExprBinary {
    lhs: ExprIdent {
        name: "ord",
    },
    op: LessThan,
    rhs: ExprBinary {
        lhs: ExprBinary {
            lhs: ExprIdent {
                name: "add0",
            },
            op: Add,
            rhs: ExprIdent {
                name: "add0",
            },
        },
        op: Add,
        rhs: ExprIdent {
            name: "add1",
        },
    },
}
ExprBinary {
    lhs: ExprIdent {
        name: "add",
    },
    op: Add,
    rhs: ExprBinary {
        lhs: ExprBinary {
            lhs: ExprIdent {
                name: "mul0",
            },
            op: Multiply,
            rhs: ExprIdent {
                name: "mul0",
            },
        },
        op: Multiply,
        rhs: ExprIdent {
            name: "mul1",
        },
    },
}
ExprBinary {
    lhs: ExprIdent {
        name: "mul",
    },
    op: Multiply,
    rhs: ExprUnary {
        rhs: ExprIdent {
            name: "unary",
        },
        op: Minus,
    },
}
ExprUnary {
    rhs: ExprCall {
        callee: ExprIdent {
            name: "call",
        },
        args: [],
    },
    op: Minus,
}
